servers/rendering/shader_preprocessor.cpp

```diff
@@ -0,0 +1,1317 @@
+	int line = p_tokenizer->get_line();
+
+	String body = tokens_to_string(p_tokenizer->advance('\n')).strip_edges();
+	if (body.is_empty()) {
+		set_error(RTR("Missing condition."), line);
+		return;
+	}
+
+	Error error = expand_macros(body, line, body);
+	if (error != OK) {
+		return;
+	}
+
+	Expression expression;
+	Vector<String> names;
+	error = expression.parse(body, names);
+	if (error != OK) {
+		set_error(expression.get_error_text(), line);
+		return;
+	}
+
+	Variant v = expression.execute(Array(), nullptr, false);
+	if (v.get_type() == Variant::NIL) {
+		set_error(RTR("Condition evaluation error."), line);
+		return;
+	}
+
+	bool success = v.booleanize();
+	start_branch_condition(p_tokenizer, success);
+}
+
+void ShaderPreprocessor::process_ifdef(PreprocessorTokenizer *p_tokenizer) {
+	const int line = p_tokenizer->get_line();
+
+	String label = p_tokenizer->get_identifier();
+	if (label.is_empty()) {
+		set_error(RTR("Invalid macro name."), line);
+		return;
+	}
+
+	p_tokenizer->skip_whitespace();
+	if (!is_char_end(p_tokenizer->peek())) {
+		set_error(RTR("Invalid ifdef."), line);
+		return;
+	}
+	p_tokenizer->advance('\n');
+
+	bool success = state->defines.has(label);
+	start_branch_condition(p_tokenizer, success);
+}
+
+void ShaderPreprocessor::process_ifndef(PreprocessorTokenizer *p_tokenizer) {
+	const int line = p_tokenizer->get_line();
+
+	String label = p_tokenizer->get_identifier();
+	if (label.is_empty()) {
+		set_error(RTR("Invalid macro name."), line);
+		return;
+	}
+
+	p_tokenizer->skip_whitespace();
+	if (!is_char_end(p_tokenizer->peek())) {
+		set_error(RTR("Invalid ifndef."), line);
+		return;
+	}
+	p_tokenizer->advance('\n');
+
+	bool success = !state->defines.has(label);
+	start_branch_condition(p_tokenizer, success);
+}
+
+void ShaderPreprocessor::process_include(PreprocessorTokenizer *p_tokenizer) {
+	const int line = p_tokenizer->get_line();
+
+	p_tokenizer->advance('"');
+	String path = tokens_to_string(p_tokenizer->advance('"'));
+	path = path.substr(0, path.length() - 1);
+	p_tokenizer->skip_whitespace();
+
+	if (path.is_empty() || !is_char_end(p_tokenizer->peek())) {
+		set_error(RTR("Invalid path."), line);
+		return;
+	}
+
+	Ref<Resource> res = ResourceLoader::load(path);
+	if (res.is_null()) {
+		set_error(RTR("Shader include load failed. Does the shader exist? Is there a cyclic dependency?"), line);
+		return;
+	}
+
+	Ref<Shader> shader = Object::cast_to<Shader>(*res);
+	Ref<ShaderInclude> shader_inc = Object::cast_to<ShaderInclude>(*res);
+	if (shader.is_null() && shader_inc.is_null()) {
+		set_error(RTR("Shader include resource type is wrong."), line);
+		return;
+	}
+	bool is_shader = shader.is_valid();
+
+	String included = is_shader ? shader->get_code() : shader_inc->get_code();
+	if (!included.is_empty()) {
+		uint64_t code_hash = included.hash64();
+		if (state->cyclic_include_hashes.find(code_hash)) {
+			set_error(RTR("Cyclic include found."), line);
+			return;
+		}
+	}
+
+	if (is_shader) {
```

Comment:
I have to say this code ended up being pretty simple in the end, so maybe you could still leave it be for include files (allowing to optionally specify shader_type if you want to edit the thing standalone?)